### ISSUE TITLE
Trashbin Deletion: Guarantee Deletion Date and Retention Obligation

### ISSUE THREAD

**Issue by MTRichards** [owncloud/core#16300]
As an ownCloud administrator, I want to set the trashbin file expiration policy so that I can better manage the disk space of users within ownCloud.

Acceptance Criteria:
- The current trashbin retention  obligation and trashbin auto expiration would be merged into one setting, with this added as well
- A modified line in the sample.config.php would highlight how this works
- The setting is now just called trashbin retention obligation
- The description is "If the trash bin app is enabled (default), this is the minimum and / or maximum number of days a file is kept in the trash before it is deleted. Setting can be number of days, auto or disabled"
- Syntax is something like 
      - 'trashbin_retention_obligation' => min 30, max 31
      - OR 'trashbin_retention_obligation' => min 30,auto
      - OR 'trashbin_retention_obligation' => min 30
      - OR 'trashbin_retention_obligation' => auto, max 30
      - OR 'trashbin_retention_obligation' => auto
      - OR 'trashbin_retention_obligation' => disabled (or just comment out the line in the config file)
- If Max is less than min, the min setting wins - so a file is always kept at least as long as the min, but in this case is immediately deleted as soon as the min is met
- If min is Auto, then the file is only kept as long as the space saving algorithm decides, up to a max of the max setting
- if max is auto, then the file is kept as long as the space saving algorithm says, and perhaps may never be actually deleted (this is the current behavior that we are adjusting)
- a setting of disabled (or a commented out config file line item) would remove this and all files would be kept forever
- if only a min value is specified, then really this is the same as no expiration or disabled
- Migration would interpret current settings as follows:
     - an auto expiration set today with no obligation is migrated to a setting of auto
     - a retention obligation set today with auto delete enabled is migrated to min today, auto
     - a retention obligation set today with auto delete disabled is migrated to min today

**Comment by cdamken** [comment]
:+1: 

BTW: https://github.com/owncloud/core/issues/8109

**Comment by MTRichards** [comment]
Slated for 8.2.

**Comment by VicDeo** [comment]
@MTRichards @DeepDiver1975 
There is an option already for disabling autoexpiration.
Should we dispose 
`'trashbin_auto_expire' => false` 
in favor of 
`'trashbin_retention_obligation' => 'disabled'` ?

**Comment by MTRichards** [comment]
:)

**Comment by VicDeo** [comment]
Some remarks.
> OR 'trashbin_retention_obligation' => disabled (or just comment out the line in the config file)

I assume that default value for new installations is `auto`, In that case commenting the whole line reverts to `auto` and not disables retention obligation. The part in parentheses is doable but it goes against the concept 'default if unset' that we  have for other settings currently.

Another thing about skipping `min` and `max` literals. I mean using
`30, 31` instead of `min 30, max 31`
`30, auto`  instead of `min 30, auto`
etc

So it can be  `auto`, `disabled` or a pair of two values and it always stands for `min, max` in the latter case. 
More Ordnung is always better.

**Comment by karlitschek** [comment]
Yes. The default should be the 8.1 behavior :-)

**Comment by MTRichards** [comment]
>Yes. The default should be the 8.1 behavior :-) 

yes, sorry if that was confusing...

**Comment by MTRichards** [comment]
>Another thing about skipping min and max literals. I mean using
30, 31 instead of min 30, max 31
30, auto instead of min 30, auto
etc

Ok, as long as the sample.config.php has this in it described properly, all is good.

**Comment by DeepDiver1975** [comment]
@MTRichards is it expected that the rules are trigger by a cron job? 
If yes this requires a followup PR of #18065

**Comment by MTRichards** [comment]
Is there any way to take care of this without a cron job and still get deletion at a specified time?

**Comment by DeepDiver1975** [comment]
> Is there any way to take care of this without a cron job and still get deletion at a specified time?

As of today the expiry will only be triggered if a user is active on the instance.
As soon as a user is not active the expiry code will not be triggered - this was a complain from at least one of the users/customers afaik

**Comment by MTRichards** [comment]
> As of today the expiry will only be triggered if a user is active on the instance.
As soon as a user is not active the expiry code will not be triggered - this was a complain from at least one of the users/customers afaik

yes, this is what this feature is trying to fix - so it sounds like cron is required to fix it properly with a guaranteed delete.

**Comment by DeepDiver1975** [comment]
> yes, this is what this feature is trying to fix - so it sounds like cron is required to fix it properly with a guaranteed delete.

same applies to version expiry as I assume ...

**Comment by MTRichards** [comment]
yes please.

**Comment by DeepDiver1975** [comment]
@VicDeo please implement a cron based expiry once #18065 is merged - THX

**Comment by SergioBertolinSG** [comment]
@MTRichards  I need a clarification in a requirement, what should happen if you put 
```
'trashbin_retention_obligation' => 'auto,30',
```
?
Files should be removed after 30 days even if the user has enough free space quota to keep them?

**Comment by PVince81** [comment]
Seems everything was merged ? Is there anything left for this task @VicDeo ?

**Comment by DeepDiver1975** [comment]
implemented -> close

**Comment by rperezb** [comment]
And QA done by @SergioBertolinSG
